File: 29-DivideTwoIntegers.cpp
/* 
    Time Complexity                              :  O(log N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

class Solution {

int divide(int divisor, int dividend) {

    if(divisor == 0) return INT_MAX;
    if(divisor == -1 and dividend == INT_MIN) return INT_MAX;

    long ldivisor = labs(divisor);
    long ldividend = labs(dividend);
    bool sign = (divisor > 0) ^ (dividend > 0);
    int ans = 0;
    while(divisor <= dividend) {
        int shift = 0;
        while(dividend >= (divisor << shift)) {
            shift++;
        }
        shift--;
        ans += (1 << shift);
        dividend -= (divisor << shift);  
    }

    return sign == true ? -ans : ans;
}

};

File: 29. Divide Two Integers.py
#Time Complexity:O(log n)
#Space Complexity:O(n)

class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        div=dividend
        divi=divisor
        dividend=abs(dividend)
        divisor=abs(divisor)
        result=0
        while dividend>=divisor:
            shift=0
            while dividend>=divisor<<shift:
                shift=shift+1
            shift=shift-1
            result=result+(1<<shift)
            dividend=dividend-(divisor<<shift)
        if div<0 and divi>0:
            return int(-result)
        if div>0 and divi<0:
            return int(-result)
        return min(2147483647,int(result))
